src/apis/web3api.py

```diff
@@ -0,0 +1,215 @@
+"""
+from os import getenv
+from typing import Dict, List, Tuple, Any, Optional
+from fractions import Fraction
+from dotenv import load_dotenv
+from web3 import Web3
+from web3.types import TxData, TxReceipt
+from eth_typing import Address, HexStr
+from hexbytes import HexBytes
+from contracts.gpv2_settlement import gpv2_settlement as gpv2Abi
+from src.trades import Trade, OrderData, OrderExecution
+from src.helper_functions import get_logger
+from src.constants import (
+    ADDRESS,
+)
+
+
+class Web3API:
+    """
+    Class for fetching data from a Web3 API.
+    """
+
+    def __init__(self):
+        load_dotenv()
+        infura_key = getenv("INFURA_KEY")
+        self.url = f"https://mainnet.infura.io/v3/{infura_key}"
+        self.web_3 = Web3(Web3.HTTPProvider(self.url))
+        self.contract = self.web_3.eth.contract(
+            address=Address(HexBytes(ADDRESS)), abi=gpv2Abi
+        )
+        self.logger = get_logger()
+
+    def get_current_block_number(self) -> Optional[int]:
+        """
+        Function that returns the current block number
+        """
+        try:
+            return int(self.web_3.eth.block_number)
+        except ValueError as err:
+            self.logger.error("Error while fetching block number: %s", err)
+            return None
+
+    def get_tx_hashes_by_block(self, start_block: int, end_block: int) -> List[str]:
+        """
+        Function filters hashes by contract address, and block ranges
+        """
+        filter_criteria = {
+            "fromBlock": int(start_block),
+            "toBlock": int(end_block),
+            "address": ADDRESS,
+            "topics": [
+                "0xa07a543ab8a018198e99ca0184c93fe9050a79400a0a723441f84de1d972cc17"
+            ],
+        }
+
+        try:
+            transactions = self.web_3.eth.filter(filter_criteria).get_all_entries()  # type: ignore
+        except ValueError as except_err:
+            self.logger.error(
+                "ValueError while fetching hashes: %s.",
+                str(except_err),
+            )
+            transactions = []
+
+        settlement_hashes_list = []
+        # transactions may have repeating hashes, since even event logs are filtered
+        # therefore, check if hash has already been added to the list
+        for transaction in transactions:
+            tx_hash = (transaction["transactionHash"]).hex()
+            if tx_hash not in settlement_hashes_list:
+                settlement_hashes_list.append(tx_hash)
+        return settlement_hashes_list
+
+    def get_transaction(self, tx_hash: str) -> Optional[TxData]:
+        """
+        Takes settlement hash as input, returns transaction data.
+        """
+        try:
+            transaction = self.web_3.eth.get_transaction(HexStr(tx_hash))
+        except ValueError as err:
+            self.logger.debug(
+                "Error while fetching transaction: %s.",
+                str(err),
```

Comment:
This type of string formatting is also a bit old school. Check out "f-strings": https://realpython.com/python-f-strings/

```suggestion
                f"Error while fetching transaction: {err}"
```

Reply:
My version of pylint complains if I make this an f-string. Only if I hack around pylint by first creating the message and passing the resulting string to the logger, does it not complain. I did not change it for now, but can do so later.